### DISCUSSION TITLE
'FindCollidingPairs' in Broadphase

### DISCUSSION THREAD

**Discussion by lmtss** [jrouwe/JoltPhysics#257] [Q&A] [answered]
It looks like Jolt's solution is to have each object traverse the tree independently, rather than a tree-tree intersection test (Simultaneous Traversal).
I'm a newbie and I'm learning BVH based collision detection, in my opinion using tree-tree test will be faster than independent traversal (reduce repetitive traversal consumption), why jolt chose independent traversal?

**Comment by jrouwe** [comment, marked as answer]
I chose to do active objects vs tree because:

1. It's simple
2. Usually there are not that many active objects compared to the total amount of objects
3. It's trivially parallelizable (all cores will take a subset of the active objects array and collide against the tree)
4. Active objects can be created during collision detection from multiple threads at the same time (an active object colliding with a sleeping object will make it active which in turn can activate another body), that complicates maintaining a tree of them
5. If you put active objects in a tree, you will need to build and/or maintain that tree. Building a tree every frame means you can start the collision detection process only after you've built the tree, updating it as things change means overhead when moving bodies around (take a look at my GDC talk - I want to move bodies around from multiple threads at the same time which complicates things)
6. In the end, the broad phase is usually not the bottleneck (the narrow phase is).

Regards,

Jorrit.

**Reply by lmtss** [reply]
Thanks, that's a novel idea for me
